README.md
# Telnyx

Elixir client for the Telnyx API (v2).

Currently supports only `Telnyx.Messages` and `Telnyx.MessagingProfiles`, but will eventually support all v2 API calls at https://developers.telnyx.com/docs/api/v2/overview.

The bulk of documentation and examples are attached to the functions. To view, run:

```elixir
iex> h Telnyx.Messages.retrieve
```

Or look in the docs at [https://hexdocs.pm/telnyx](https://hexdocs.pm/telnyx).

## Installation

Add `telnyx` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:telnyx, "~> 0.2.1"}
  ]
end
```

## Usage

```elixir
api_key = "YOUR_API_KEY"

{:ok, data} =
  %{
    from: "+18665552368", # Your Telnyx number
    to: "+18445552367",
    text: "Hello, World!"
  }
  |> Telnyx.Messages.create(api_key)
```

## Additional Information

A concious choice was made to not force the `api_key` to live in the application environment. I've worked on projects where multiple accounts (and subsequently API keys) were being accessed, and had a bad time with libraries forcing me to put a single API key in global config.

Effort was taken to provide an API experience consistent with the officially supported Telnyx libraries for other programming languages. To me, the benefit is that as you're reading the official docs when you need to explore something more deeply, you have more equivalent code examples to look at.

## Contributing

Please let me know about any problems or suggestions on the issues page!



CHANGELOG.md
# Changelog

## 0.3.0 (2020-06-29)

*(sort of) Breaking Change:* You can now swap out Mojito, the default API client, with any you'd like. Documentation is in the Telnyx.Client module. Significantly, since Mojito is now an optional dependency, you have to add it to your deps like `{:mojito, "~> 0.7.0"}` so the project will compile.

Also adds endpoints for messaging phone numbers.

## 0.2.1 (2020-06-28)

Bumps the version of Mojito to 7.x to take advantage of [this bug-fix](https://github.com/appcues/mojito/pull/59). Previously, some requests to Telnyx were intermittently timing-out, resulting in poor user experiences.

## 0.2.0 (2019-12-15)

*Breaking Changes:*
All API calls to `Telnyx.Message` should now go to `Telnyx.Messages`, and the order of arguments has switched. The `api_key` arg now comes after the main argument for better piping. For example:

```elixir
# Old way:
api_key = "test"

Telnyx.Message.retrieve(api_key, "uuid")

# New way: 
api_key = "test"

"uuid"
|> Telnyx.Messages.retrieve(api_key)

```

Changed the module name to `Telnyx.Messages` (from `Message`) to better reflect elixir conventions.

Switched order of arguments so the primary argument is first for better piping instead of `api_key` being first.

Added `Telnyx.MessagingProfiles` endpoints.

Deprecated `create_alphanumeric/2` since it was removed from the v2 Telnyx docs.

Updated Mojito dependency to `0.6.0` so that it uses Mint `1.0.0`.

## 0.1.2 (2019-08-24)

If the response didn't have a 200 status, the error message was raising trying to use Access behavior on a struct. This fixes it and adds a couple of tests.

## 0.1.1 (2019-08-20)

Build was failing when imported into another project, because global config was being used to determine test environment. Global config was replaced by local config.

## 0.1.0 (2019-08-18)

Initial (alpha) release, supporting only the `Telnyx.Message.create/2` action. Tests are passing and a lot of structure has been built.

lib/telnyx/messages.ex
defmodule Telnyx.Messages do
  @moduledoc """
  Send a message with `create/2`, and retrieve a message with `retrieve/2`.
  """

  alias Telnyx.Client

  @doc """
  Sends a message.

  ## Examples

    ```
    api_key = "YOUR_API_KEY"

    %{
      from: "+18665552368", # Your Telnyx number
      to: "+18445552367",
      text: "Hello, World!"
    }
    |> Telnyx.Messages.create(api_key)
    ```
    Example response:
    ```
    {:ok,
      %{
        "carrier" => "Verizon",
        "cost" => %{"amount" => "1.23", "currency" => "USD"},
        "created_at" => "2019-01-23T18:10:02.574Z",
        "direction" => "outbound",
        "errors" => [],
        "from" => "+18665552368",
        "id" => "3fa85f64-5717-4562-b3fc-2c963f66afa6",
        "line_type" => "Wireless",
        "parts" => 1,
        "record_type" => "message",
        "text" => "Hello world",
        "to" => [
          %{
            "address" => "+18665550001",
            "status" => "queued",
            "updated_at" => "2019-01-23T18:10:02.574Z"
          }
        ],
        "type" => "sms",
        "updated_at" => "2019-01-23T18:10:02.574Z",
        "use_profile_webhooks" => false,
        "valid_until" => "2019-01-23T18:10:02.574Z",
        "webhook_failover_url" => "https://www.example.com/callbacks",
        "webhook_url" => "https://www.example.com/hooks"
      }
    }
    ```

  See https://developers.telnyx.com/docs/api/v2/messaging/Messages
  """
  @spec create(map, String.t()) :: {:ok, map} | {:error, %Telnyx.Error{}}
  def create(params = %{}, api_key) do
    Client.post(api_key, params, "/messages")
  end

  @doc """
  Same as `create/2` but specifically for sending long-code messages.

  See the documentation for `create/2` for example.

  `create/2` supports long-code already, if the `from` param is a long code number.

  See https://developers.telnyx.com/docs/api/v2/messaging/Messages#createLongCodeMessage
  """
  @spec create_long_code(map, String.t()) :: {:ok, map} | {:error, %Telnyx.Error{}}
  def create_long_code(params = %{}, api_key) do
    Client.post(api_key, params, "/messages/long_code")
  end

  @doc """
  Same as `create/2` but specifically for sending messages from a number pool.
  The option pool must be turned on in the messaging profile.

  `create/2` supports sending from a number pool already with the same params (ensuring that `messaging_profile_id` is included, but omitting the `from`).

  ## Examples

    ```
    api_key = "YOUR_API_KEY"

    %{
      messaging_profile_id: "uuid",
      to: "+18445552367",
      text: "Hello, World!"
    }
    |> Telnyx.Messages.create(api_key)
    ```
    Example response:
    ```
    {:ok,
      %{
        "carrier" => "Verizon",
        "cost" => %{"amount" => "1.23", "currency" => "USD"},
        "created_at" => "2019-01-23T18:10:02.574Z",
        "direction" => "outbound",
        "errors" => [],
        "from" => "+18665552368",
        "id" => "3fa85f64-5717-4562-b3fc-2c963f66afa6",
        "line_type" => "Wireless",
        "parts" => 1,
        "record_type" => "message",
        "text" => "Hello world",
        "to" => [
          %{
            "address" => "+18665550001",
            "status" => "queued",
            "updated_at" => "2019-01-23T18:10:02.574Z"
          }
        ],
        "type" => "sms",
        "updated_at" => "2019-01-23T18:10:02.574Z",
        "use_profile_webhooks" => false,
        "valid_until" => "2019-01-23T18:10:02.574Z",
        "webhook_failover_url" => "https://www.example.com/callbacks",
        "webhook_url" => "https://www.example.com/hooks"
      }
    }
    ```

  See https://developers.telnyx.com/docs/api/v2/messaging/Messages#createNumberPoolMessage
  """
  @spec create_from_number_pool(map, String.t()) :: {:ok, map} | {:error, %Telnyx.Error{}}
  def create_from_number_pool(params = %{}, api_key) do
    Client.post(api_key, params, "/messages/number_pool")
  end

  @doc """
  Same as `create/2` but specifically for sending short-code messages.

  `create/2` supports short-code already with the same params.

  See the `create/2` documentation for example.

  See https://developers.telnyx.com/docs/api/v2/messaging/Messages#createShortCodeMessage
  """
  @spec create_short_code(map, String.t()) :: {:ok, map} | {:error, %Telnyx.Error{}}
  def create_short_code(params = %{}, api_key) do
    Client.post(api_key, params, "/messages/short_code")
  end

  @doc """
  Retrieves a message.

  ## Examples

    ```
    api_key = "YOUR_API_KEY"

    Telnyx.Messages.retrieve(api_key, "uuid")
    ```

    Example response:
    ```
    {:ok,
      %{
        "carrier" => "Verizon",
        "cost" => %{"amount" => "1.23", "currency" => "USD"},
        "created_at" => "2019-01-23T18:10:02.574Z",
        "direction" => "outbound",
        "errors" => [],
        "from" => "+18445550001",
        "id" => "uuid",
        "line_type" => "Wireless",
        "parts" => 1,
        "record_type" => "message",
        "text" => "Hello, World!",
        "to" => [
          %{
            "address" => "+18665550001",
            "status" => "queued",
            "updated_at" => "2019-01-23T18:10:02.574Z"
          }
        ],
        "type" => "sms",
        "updated_at" => "2019-01-23T18:10:02.574Z",
        "use_profile_webhooks" => false,
        "valid_until" => "2019-01-23T18:10:02.574Z",
        "webhook_failover_url" => "https://www.example.com/callbacks",
        "webhook_url" => "https://www.example.com/hooks"
      }
    }
    ```

  See https://developers.telnyx.com/docs/api/v2/messaging/Messages#retrieveMessage
  """
  @spec retrieve(String.t(), String.t()) :: {:ok, map} | {:error, %Telnyx.Error{}}
  def retrieve(uuid, api_key), do: Client.get(api_key, "/messages/#{uuid}")

  @doc """
  This call disappeared out of the v2 docs, so it has been deprecated.

  You can use an alphanumeric key instead of a phone number in `create/2`.

  See `create/2` documentation for example.
  """
  @deprecated "Use create/2 instead"
  @spec create_alphanumeric(map, String.t()) :: {:ok, map} | {:error, %Telnyx.Error{}}
  def create_alphanumeric(params = %{}, api_key) do
    Client.post(api_key, params, "/messages/alphanumeric_sender_id")
  end
end
